### PR TITLE
Remove deprecated `LightningModule.on_post_move_to_device`

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Part of #12521.

### Does your PR introduce any breaking changes? If yes, please list them.
Removes `LightningModule.on_post_move_to_device`.

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @kaushikb11 @rohitgr7 @borda @carmocca @justusschock @awaelchli @ananthsub @ninginthecloud @jjenniferdai

### CODE REVIEW

File: docs/source-pytorch/accelerators/tpu_advanced.rst [hunk 1]
```diff
@@ -12,11 +12,10 @@ Weight Tying/Sharing is a technique where in the module weights are shared among
...
 
-PyTorch XLA requires these weights to be tied/shared after moving the model
-to the TPU device. To support this requirement Lightning provides a model hook which is
-called after the model is moved to the device. Any weights that require to be tied should
-be done in the `on_post_move_to_device` model hook. This will ensure that the weights
-among the modules are shared and not copied.
+PyTorch XLA requires these weights to be tied/shared after moving the model to the XLA device.
+To support this requirement, Lightning automatically finds these weights and ties them after
+the modules are moved to the XLA device under the hood. It will ensure that the weights among
+the modules are shared but not copied independently.
```
Comment (akihironitta) [inline docs/source-pytorch/accelerators/tpu_advanced.rst:18]: @kaushikb11 Could you review this change in the doc page? My understanding is that #8555 suggested tying the weights without the need of user overriding `on_post_move_to_device`, and #9525 addressed the issue without updating the docs.

**Review by awaelchli — APPROVED** [review]

**Review by kaushikb11 — APPROVED** [review]
LGTM!

**Review by Borda — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Comment by andreimargeloiu** [comment]
Is there a replacement for ```LightningModule.on_post_move_to_device``` ?

**Comment by carmocca** [comment]
@andreimargeloiu this hook was misleading as it was only called when using the tpu spawn strategy. It's likely different from what you are looking for. But to answer you, the next hook to run would be `on_fit_start`

If you want a hook related to moving a batch to device, you can use one of `on_before_batch_transfer`, `transfer_batch_to_device`, `on_after_batch_transfer`

**Pull request merged**